src/http/one/Tokenizer.cc

```diff
@@ -78,32 +60,41 @@ Http::One::Tokenizer::qdText(SBuf &returnedToken, const bool http1p0)
              */
             static const CharacterSet qPairChars = CharacterSet::HTAB + CharacterSet::SP + CharacterSet::VCHAR + CharacterSet::OBSTEXT;
             SBuf escaped;
-            if (!prefix(escaped, qPairChars, 1)) {
-                returnedToken.clear();
-                restoreLastCheckpoint();
-                return false;
-            }
+            if (!tok.prefix(escaped, qPairChars, 1))
+                throw TexcHere("invalid escaped characters");
+
             returnedToken.append(escaped);
             continue;
+        } else if (tok.skip('"')) {
+            return true;
+        } else if (tok.atEnd()) {
+            break;
+        }
+        throw TexcHere(ToSBuf("invalid bytes for set ", tokenChars.name));
+    }
 
-        } else if (skip('"')) {
-            break; // done
+    return false; // need more data
+}
 
-        } else if (atEnd()) {
-            // need more data
-            returnedToken.clear();
-            restoreLastCheckpoint();
+bool
+Http::One::tokenOrQuotedString(Parser::Tokenizer &tok, SBuf &returnedToken, const bool tokenPrefixResult, const bool http1p0)
+{
+    if (tok.skip('"')) {
+        const auto savedTok = tok;
+        SBuf parsedToken;
+        if (!quotedString(tok, parsedToken, http1p0)) {
+            tok = savedTok;
             return false;
         }
-
-        // else, we have an error
-        debugs(24, 8, "invalid bytes for set " << tokenChars.name);
-        returnedToken.clear();
-        restoreLastCheckpoint();
-        return false;
+        returnedToken = parsedToken;
+        return true;
     }
 
-    // found the whole string
+    if (!tok.prefix(returnedToken, CharacterSet::TCHAR))
+        return false;
+    if (tok.atEnd() && !tokenPrefixResult)
```

Comment:
Please use tokenPrefixResult directly as intended/defined:

```suggestion
    if (tok.atEnd())
        return tokenPrefixResult;
```

Reply:
Done.